### DISCUSSION TITLE
Alpha: Payments

### DISCUSSION THREAD

**Discussion by 0x4007** [ubiquity/ubiquibot#68] [Specifications]
- [x] Registering wallet addresses and associating it to each GitHub username
  - The Bounty Hunter should be able to set their wallet address with the following command: 
    - > /wallet pavlovcik.eth
  - If a user is assigned to an issue, the bot should immediately check if the author has a registered address. If they do not, then it should comment with instructions; something like 
    - > @user your payment wallet address is not currently registered. Please register your payment wallet address by commenting your address in the following format "/wallet ubq.eth" or "/wallet 0xefC0e701A824943b469a694aC564Aa1efF7Ab7dd" this wallet address will be used to receive payment upon the successful completion of this bounty. 
      - If the bounty hunter registers their payment address as `ubq.eth` or `0xefC0e701A824943b469a694aC564Aa1efF7Ab7dd` (they misunderstand the example) throw an error as a comment and say something like: 
      - > Failed to register your wallet address. You must include the address to your wallet, not `ubq.eth` or `0xefC0e701A824943b469a694aC564Aa1efF7Ab7dd`
  - If the user successfully registers their wallet address, the bot should provide a confirmation:
    - > @user's payment wallet address is now set to `pavlovcik.eth`.

- [x] Paying Bounties
  - Please see the code [here](https://github.com/ubiquity/generate-permit) to learn how we generate payment permits.
  - The bot will need to know the bounty hunter's address to generate a permit. 
  - When 1. the bounty is complete and 2. the bot has the user's wallet address, then it will generate the permit and post a link.
  - Bot posts a link to our "DevPool Claims" (view and claim PERMIT) interface. 
  - > **[[ CLAIM 1 DAI ]](https://pay.ubq.fi/?claim=eyJwZXJtaXQiOnsicGVybWl0dGVkIjp7InRva2VuIjoiMHg2QjE3NTQ3NEU4OTA5NEM0NERhOThiOTU0RWVkZUFDNDk1MjcxZDBGIiwiYW1vdW50IjoiMTAwMDAwMDAwMDAwMDAwMDAwMCJ9LCJub25jZSI6IjYxMDk3Nzk0MzA3NTU3NzU4ODUxNDc3Njc5NDc4Njg4MjkxNDQwMDIwOTkxNTEwNTU4OTM4NzkzMDE5NzExMDg0OTE4NTA2NjQ5NjA0IiwiZGVhZGxpbmUiOiIxMTU3OTIwODkyMzczMTYxOTU0MjM1NzA5ODUwMDg2ODc5MDc4NTMyNjk5ODQ2NjU2NDA1NjQwMzk0NTc1ODQwMDc5MTMxMjk2Mzk5MzUifSwidHJhbnNmZXJEZXRhaWxzIjp7InRvIjoiMHg0MDA3Q0UyMDgzYzdGM0UxODA5N2FlQjNBMzliYjhlQzE0OWEzNDFkIiwicmVxdWVzdGVkQW1vdW50IjoiMTAwMDAwMDAwMDAwMDAwMDAwMCJ9LCJvd25lciI6IjB4Zjg3Y2E0NTgzQzc5MjIxMmU1MjcyMGQxMjdFN0UwQTM4QjgxOGFEMSIsInNpZ25hdHVyZSI6IjB4MmYwMTNlYWM3YzIyNDkxYTkwZjU1OTIxYmFjMTkwMWZmZTdiNDk3OGZhZWQ5M2Y0OWJhZGU1YTBlMTEzYjUxNjcxZThmY2Q4MzNiYTFmNDlkZDJhOTcxYzYzNjgxNjc5ODE5NGVhZWExMjJmMzRjODgyNWNmN2Q1NTBjMmFlYzExYiJ9)**

- [x] Registering Payments Made 
The research for this is not completed. The idea is associating payments with the GitHub username (for later accounting/auditing purposes.)
  - This can look for Etherscan transactions posted by Bounty masters.
  - The bot should confirm the logging of this transaction via a comment.

**Comment by 0x4007** [comment]
### Paying Bounties V2

We should eventually consider offering a service to pay the bounties on behalf of hunters because we can charge a profit for it. The following is taken from the original spec. Consider it a working draft.

  - This is [spec'd out](https://github.com/ubiquity/bounty-bot/issues/13) to allow for automatic payments OR queuing payments via Gnosis Safe. 
  - This should check if the Bounty Hunter has a registered address in the database. If they don't then the bot should throw an error in the comments repeating the original instructions on how to set the wallet, and then also add extra detail letting them know that the bot will not be handling the bounty automatically, so we don't have to worry about state. 
  - > Please register your payment wallet address by commenting your Etherscan link in the following format "/wallet https://etherscan.io/address/0xefC0e701A824943b469a694aC564Aa1efF7Ab7dd"
  - > I am no longer monitoring this issue because it is closed. Please request payment from one of the Bounty Masters after you have registered your wallet.
  - Should also check who recruited the Bounty Hunter and give them their commission.

**Comment by sergfeldman** [comment]
### Pay with Vault

**Goals**

- Increase income for bounty hunters
- Decrease cost of bounty resolution for crypto projects
- Popularize Ubiquity Yield aggregator

**Proposal**

Partner crypto project deposits bounty budget into the Ubiquity Vault.

While bounties are being developed the partner project gets yield generated by the Vault, so the bounty implementation is performed with kind of discount from regular price.

For regular bounty hunters, part of the payment (or the entire payment) is paid not in stablecoins, but in the Ubiquity Vault LP tokens.

If they want, they can instantly convert LP tokens into stablecoins.

If they can wait, they will receive an additional yield generated by the Ubiquity Vault.

**Yield multiplier** 

The Vault yield multiplier can be based on the efficiency level of the developer as a DevPool bounty hunter.

For example,

- Noob (level 1) - multiplier 1 (for example 10% APY)
- Master (level 2) - multiplier 1.5 (for example 15% APY)
- Chad (level 3) - multiplier 2 (for example 20% APY)

**Benefits**

It will stimulate regular bounty hunters to maintain the high level of their efficiency as bounty hunters in the DevPool.

It will help building more stable relationships with partner crypto projects.

This way DevPool and Yield aggregator products will help each other grow.

**Reply by sergfeldman** [reply]
A simple case - DAI:uAD LP or ETH:USDC LP 

A complex case - Ubiquity Vault with Bring Your Yield Source logic, where partner projects provided their yield sources
